<!doctype html>
<html lang="es-es">   
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>var-node</title>
  </head>
  <script src="{{ url_for('static', filename='jquery-3.5.1.min.js') }}"></script>
  <script src="{{ url_for('static', filename='bootstrap.min.js') }}"></script>
  <link href="{{ url_for('static', filename='bootstrap.min.css') }}" rel="stylesheet" type="text/css" media="all" >
  <link href="{{ url_for('static', filename='custom.css') }}" rel="stylesheet" type="text/css" media="all" >
  <link href="{{ url_for('static', filename='font-awesome/css/font-awesome.min.css')}}" rel='stylesheet'type="text/css" media="all">
  <script type="text/javascript">

    var variant_id_data = JSON.parse('{{ variant_id_data|tojson }}');
    var results = JSON.parse('{{ results|tojson }}');

    console.log(variant_id_data);
    console.log(results);

    $( document ).ready(function() {

      if( ! results ){
        $("#spinner").append('<i class="fa fa-spinner fa-pulse fa-3x fa-fw"></i>');
        $(".btn-spinner").addClass('disabled');
        let delay = 1; 
        let url = "/info"; 
        setTimeout(function(){ 
            location = url; 
        },delay)
      }

      if( variant_id_data["variant_id"] ){
        $("#input-genomic-variant").val(variant_id_data["variant_id"]);
      }
      if( variant_id_data["genome"]  ){
        $("input[name=genome_radio]").val([variant_id_data["genome"]]);
      }else{
        $("input[name=genome_radio]").val(["grch37"]);
      }

      $(".btn-spinner").click( function(){
        $("#spinner").append('<i class="fa fa-spinner fa-pulse fa-3x fa-fw"></i>');
      });

      $("#submit-btn").click( function(){
        if( $("#input-genomic-variant").val() ){
          window.location.href = "/" + $("input[name=genome_radio]:checked").val() + "/" + $("#input-genomic-variant").val();
        }else{
          alert("Variant ID field is empty.");
          $("#spinner").empty();
        }
      });

      $("#to-json-btn").click( function(){
        if( $("#input-genomic-variant").val() ){
          window.location.href = "/json/" + $("input[name=genome_radio]:checked").val() + "/" + $("#input-genomic-variant").val();
        }else{
          alert("Variant ID field is empty.");
          $("#spinner").empty();
        }
      });

      $("#liftover-btn").click( function(){
        if( $("#input-genomic-variant").val() ){
          window.location.href = "/liftover/" + $("input[name=genome_radio]:checked").val() + "/" + $("#input-genomic-variant").val();
        }else{
          alert("Variant ID field is empty.");
          $("#spinner").empty();
        }
      });

      $("#reset-btn").click( function(){
        window.location.href = "/";
      });

      $("#info-btn").click( function(){
        $("#info-modal").modal("show");
      });
      

      $("#input-genomic-variant").on( "keypress", function( event ) {
        if ( event.which == 13 ) {
          $("#submit-btn").click();
        }
      });
      $( ".genome-radio" ).on( "keypress", function( event ) {
        if ( event.which == 13 ) {
          $("#submit-btn").click();
        }
      });

      if( variant_id_data["variant_id"] ){
        if(  variant_id_data["validation"] == "OK" ){
          $('#variant-info').append(`
            <div class="panel panel-success top-buffer">
              <div class="panel-heading" data-toggle="collapse" data-target="#collapse_variant_id">
                <div class="row">
                  <div class="col-xs-11 text-left long-text"><b>Variant validation: ` + variant_id_data["variant_id"] + ` (` + variant_id_data["genome"] + `)</b></div>
                  <div class="col-xs-1 text-right"><b>` + variant_id_data["validation"] + `</b></div>
                </div>
              </div>
              <div id="collapse_variant_id" class="panel-collapse collapse">
                <div class="panel-body">
                  <div class="row top-buffer">
                    <div class="col-xs-12">
                      <div id="transcripts-data"></div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          `);
          var transcripts_html = '<ul class="list-group">';
          for (var key in variant_id_data["results"] ){
            transcripts_html = transcripts_html + '<li class="list-group-item"><div class="row"><div class="col-xs-8 text-left long-text">' + variant_id_data["results"][key]["internal_id"] + '</div><div class="col-xs-3 text-left long-text">' + variant_id_data["results"][key]["consequence"] + '</div><div class="col-xs-1 text-right">' + variant_id_data["results"][key]["impact"] + '</div></div>' + '</li>'
          }
          transcripts_html = transcripts_html + '</ul>';
          $('#transcripts-data').append(transcripts_html);
        }else{
          $('#variant-info').append(`
            <div class="panel panel-danger">
              <div class="panel-heading">
                <div class="row">
                  <div class="col-xs-9 text-left long-text"><b>Variant validation: ` + variant_id_data["variant_id"] + `</b></div>
                  <div class="col-xs-3 text-right"><b>` + variant_id_data["validation"] + `</b></div>
                </div>
              </div>
            </div>
          `);
        }
      }
      if( results ){
        for (var node_number in results ){
          if( results[node_number]["error"] == "Ready" ){
            $('#nodes-data').append(`
              <div class="panel panel-success">
                <div class="panel-heading">
                  <div class="row">
                    <div class="col-xs-6 text-left"><b>` + results[node_number]["node_name"] + ` (` + results[node_number]["node_host"] + `:` + results[node_number]["node_port"] + `)</b></div>
                    <div class="col-xs-6 text-right"><b>Avalaible genomes: ` + results[node_number]["database_genomes"]+ `</b></div>
                  </div>
                </div>
              </div>
            `);
          }else if( results[node_number]["error"] == "OK" ){
            $('#nodes-data').append(`
              <div class="panel-group">
                <div class="panel panel-success">
                  <div class="panel-heading" data-toggle="collapse" data-target="#collapse` + node_number + `">
                    <div class="row">
                      <div class="col-xs-6 text-left"><b>` + results[node_number]["node_name"] + ` (` + results[node_number]["node_host"] + `:` + results[node_number]["node_port"] + `)</b></div>
                      <div class="col-xs-6 text-right"><b>Allele count: ` + results[node_number]["allele_count"] + `</b></div>
                    </div>
                  </div>
                  <div id="collapse` + node_number + `" class="panel-collapse collapse">
                    <div class="panel-body">
                      <div class="row">
                        <div class="col-xs-2 text-right"><b>Variant ID: </b></div>
                        <div class="col-xs-10 text-left long-text">` + results[node_number]["variant_id"] + `</div>
                      </div>
                      <div class="row">
                        <div class="col-xs-2 text-right"><b>Genome: </b></div>
                        <div class="col-xs-10 text-left">` + results[node_number]["genome"] + `</div>
                      </div>
                      <div class="row">
                        <div class="col-xs-2 text-right"><b>Allele count: </b></div>
                        <div class="col-xs-1 text-left">` + results[node_number]["allele_count"] + `</div>
                        <div class="col-xs-2 text-right"><b>Heterozygotes: </b></div>
                        <div class="col-xs-1 text-left">` + results[node_number]["heterozygotes"] + `</div>
                        <div class="col-xs-2 text-right"><b>Homozygotes: </b></div>
                        <div class="col-xs-1 text-left">` + results[node_number]["homozygotes"] + `</div>
                      </div>
                      <div class="row top-buffer">
                        <div class="col-xs-12">
                          <div id="samples-data-`+node_number+`"></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            `);
            var samples_html = '<ul class="list-group">';
            samples_html = samples_html + '<li class="list-group-item"><div class="row"><div class="col-xs-1 text-left"><b>SAMPLE</b></div><div class="col-xs-1 text-left"><b>ZIGOSITY</b></div><div class="col-xs-7 text-left"><b>SAMPLE METADATA</b></div><div class="col-xs-3 text-left"><b>GENOTYPE DETAILS</b></div></div>' + '</li>';
            for (var key in results[node_number]["samples"] ){
              samples_html = samples_html + '<li class="list-group-item"><div class="row"><div class="col-xs-1 text-left">IND' + key + '</div><div class="col-xs-1 text-left">' + results[node_number]["samples"][key]["zigosity"] + '</div><div class="col-xs-7 text-left long-text">' + JSON.stringify(results[node_number]["samples"][key]["sample_details"],null,2) + '</div><div class="col-xs-3 text-left long-text">' + JSON.stringify(results[node_number]["samples"][key]["genotype_details"],null,2) + '</div></div>' + '</li>';
            }
            if( results[node_number]["samples"].length >= 50 ){
              samples_html = samples_html + '<li class="list-group-item"><div class="row"><div class="col-xs-12 text-left text-danger"><i>... more than 50 samples ...</i></div></div>' + '</li>';
            }
            samples_html = samples_html + '</ul>';
            $('#samples-data-' + node_number).append(samples_html);
          }else{
            $('#nodes-data').append(`
              <div class="panel panel-danger">
                <div class="panel-heading">
                  <div class="row">
                    <div class="col-xs-6 text-left"><b>` + results[node_number]["node_name"] + ` (` + results[node_number]["node_host"] + `:` + results[node_number]["node_port"] + `)</b></div>
                    <div class="col-xs-6 text-right"><b>` + results[node_number]["error"] + `</b></div>
                  </div>
                </div>
              </div>
            `);
          }
        }
      }
    });
  </script>
  <body>
    <div class="container-fluid">
      <div class="row top-buffer">
        <div class="col-xs-8 col-xs-offset-2 ">
          <div class="panel panel-default">
            <div class="panel-heading">
              <div class="row">
                <div class="col-xs-10 text-left">
                  <h4>var-node - {{ node_name }} - {{ network_name }}</h4>
                </div>
                <div class="col-xs-2 text-right">
                  <div class="text-right" id="spinner"></div>
                </div>
              </div>
            </div>
            <div class="panel-body">
              <div class="row">
                <div class="col-xs-2 text-right">
                  <b>Genomic Variant: </b>
                </div>
                <div class="col-xs-6 text-left">
                  <input type="text" class="form-control" id="input-genomic-variant" placeholder="chr-pos-ref-alt">
                </div>
                <div class="col-xs-4 text-left">
                  <button type="button" id="submit-btn" class="btn btn-default btn-spinner">Submit</button>
                  <button type="button" id="reset-btn" class="btn btn-default btn-spinner">Reset</button>
                  <button type="button" id="to-json-btn" class="btn btn-default btn-spinner">JSON</button>
                  <button type="button" id="liftover-btn" class="btn btn-default btn-spinner">LiftOver</button>
                  <button type="button" id="info-btn" class="btn btn-default pull-right">Info</button>                  
                </div>
              </div>
              <div class="row top-buffer">
                <div class="col-xs-2 text-right">
                  <b>Reference genome: </b>
                </div>
                <div class="col-xs-8 text-left">
                  <label class="radio-inline genome-radio"><input type="radio" name="genome_radio" value="grch37">GRCh37 / hg19</label>
                  <label class="radio-inline genome-radio"><input type="radio" name="genome_radio" value="grch38">GRCh38 / hg38</label>
                </div>
              </div>
              <div class="row top-buffer">
                <div class="col-xs-2 text-right">
                  <b>Examples: </b>
                </div>
                <div class="col-xs-8 text-left">
                  <div class="row">
                    <div class="col-xs-2">
                      GRCh37
                    </div>
                    <div class="col-xs-6">
                      CUBN(NM_001081.4):c.4675C>T:p.Pro1559Ser
                    </div>
                    <div class="col-xs-4">
                      <a href="/grch37/10-17024503-G-A" class="btn-spinner" role="button">grch37/10-17024503-G-A</a>
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-2">
                      GRCh38
                    </div>
                    <div class="col-xs-6">
                      CUBN(NM_001081.4):c.4675C>T:p.Pro1559Ser
                    </div>
                    <div class="col-xs-4">
                      <a href="/grch38/10-16982504-G-A" class="btn-spinner" role="button">grch38/10-16982504-G-A</a>
                    </div>
                  </div>                 
                </div>
              </div>
              
            </div>
          </div>
          <div id="variant-info">
          </div>
          <div id="nodes-data">
          </div>
        </div>
      </div>

      <!-- Modal -->
      <div id="info-modal" class="modal fade" tabindex="-1" role="dialog">
        <div class="modal-dialog modal-lg">
            <div class="modal-content">
                <div class="modal-body">
                  <div class="row">
                    <div class="col-xs-12">
                      <h4>Important notes</h4>
                    </div>
                  </div>
                  <div class="row top-buffer">
                    <div class="col-xs-12 text-left">
                      - only variants mapped to primary assembled chromosomes are inserted into the database:            
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-12 text-left">
                      &emsp;1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, X, Y, MT             
                    </div>
                  </div>
                  <div class="row top-buffer">
                    <div class="col-xs-12 text-left">
                      - variants with hg19 / hg38 chromosomes have been lift over their equivalent GRCh37 / GRCh38 chromosome nomenclature:           
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-12 text-left">
                      &emsp;chr1->1, chr2->2, chr3->3, chr4->4, chr5->5, chr6->6, chr7->7, chr8->8, chr9->9, chr10->10, chr11->11, chr12->12, chr13->13, chr14->14, chr15->15, chr16->16, chr17->17, chr18->18, chr19->19, chr20->20, chr21->21, chr22->22, chrX->X, chrY->Y, chrM->MT             
                    </div>
                  </div>
                  <div class="row top-buffer">
                    <div class="col-xs-12 text-left">
                      - all the variants in the database have been left-aligned and normalized, and multiallelic sites have been splited into biallelic records, using the following Ensembl primary assemblies and <code>bcftools norm</code> command line:              
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-12 text-left">
                      &emsp;<code>https://ftp.ensembl.org/pub/grch37/release-111/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.dna.primary_assembly.fa.gz</code>             
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-12 text-left">
                      &emsp;<code>https://ftp.ensembl.org/pub/release-111/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</code>             
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-12 text-left">
                      &emsp;<code>bcftools norm --fasta-ref $FASTA --multiallelics -any --check-ref wx</code>           
                    </div>
                  </div>
                  <div class="row top-buffer">
                    <div class="col-xs-12 text-left">
                      - queried variant normalization is performed on-the-fly using <code>bcftools norm</code>:           
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-12 text-left">
                      &emsp;<code>bcftools norm --fasta-ref $FASTA --check-ref wx</code>          
                    </div>
                  </div>
                  <div class="row top-buffer">
                    <div class="col-xs-12 text-left">
                      - queried variant annotation is performed on-the-fly using VEP version 111 with the following vep_cache files:           
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-12 text-left">
                      &emsp;<code>https://ftp.ensembl.org/pub/grch37/release-111/variation/indexed_vep_cache/homo_sapiens_merged_vep_111_GRCh37.tar.gz</code>             
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-12 text-left">
                      &emsp;<code >https://ftp.ensembl.org/pub/release-111/variation/indexed_vep_cache/homo_sapiens_merged_vep_111_GRCh38.tar.gz</code>             
                    </div>
                  </div>
                  <div class="row top-buffer">
                    <div class="col-xs-12 text-left">
                      - queried variant liftover is performed on-the-fly using <code>bcftools +liftover</code> with the following chain files:          
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-12 text-left">
                      &emsp;<code >https://ftp.ensembl.org/pub/assembly_mapping/homo_sapiens/GRCh37_to_GRCh38.chain.gz</code>             
                    </div>
                  </div>
                  <div class="row">
                    <div class="col-xs-12 text-left">
                      &emsp;<code >https://ftp.ensembl.org/pub/assembly_mapping/homo_sapiens/GRCh38_to_GRCh37.chain.gz</code>             
                    </div>
                  </div>
                  <div class="row top-buffer">
                    <div class="col-xs-12 text-left">
                      - variants containing * (star) or . (missing) alleles are not included in the database           
                    </div>
                  </div>
                  <div class="row top-buffer">
                    <div class="col-xs-12 text-left">
                      - a maximum of 50 random samples per node will be shown     
                    </div>
                  </div>
                </div>
            </div>
        </div>


    </div>
  </body> 
</html>

